File: banco_problemas_aplicados/nivel_intermedio/soluciones/java/Problema23.java
/*
 * Problema 23 - Inversion Espectral In-Place en Memoria
 * void invertirBuffer(int[] arr, int tam)  -> invierte in-place, solo una variable temporal.
 */
import java.util.Scanner;

public class Problema23 {

    static void invertirBuffer(int[] arr, int tam) {
        for (int i = 0; i < tam / 2; i++) {
            int aux = arr[i];
            arr[i] = arr[tam - 1 - i];
            arr[tam - 1 - i] = aux;
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Ingrese el tamano del buffer: ");
        int tam = sc.nextInt();

        int[] arr = new int[tam];
        for (int i = 0; i < tam; i++) {
            System.out.print("Dato " + i + ": ");
            arr[i] = sc.nextInt();
        }

        invertirBuffer(arr, tam);

        System.out.print("Buffer invertido: ");
        for (int i = 0; i < tam; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
        sc.close();
    }
}

File: banco_problemas_aplicados/nivel_intermedio/soluciones/c/problema23.c
/*
 * Problema 23 - Inversion Espectral In-Place en Memoria
 * void invertirBuffer(int Arr[], int tam)
 * Contexto: optimizacion de RAM. Invierte el orden del arreglo SIN vectores auxiliares
 * (solo una variable temporal), intercambiando Arr[i] con Arr[tam-1-i].
 */
#include <stdio.h>

void invertirBuffer(int arr[], int tam) {
    for (int i = 0; i < tam / 2; i++) {
        int aux = arr[i];                    /* unica variable temporal, no un arreglo */
        arr[i] = arr[tam - 1 - i];
        arr[tam - 1 - i] = aux;
    }
}

int main(void) {
    int tam;
    printf("Ingrese el tamano del buffer: ");
    scanf("%d", &tam);

    int arr[tam];
    for (int i = 0; i < tam; i++) {
        printf("Dato %d: ", i);
        scanf("%d", &arr[i]);
    }

    invertirBuffer(arr, tam);

    printf("Buffer invertido: ");
    for (int i = 0; i < tam; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
    return 0;
}
